### PR TITLE
Drop glyph names for CFF fonts, promote to CID

### PR DESCRIPTION
**Pull request by madig**
Copy-pasting of Behdad's code in https://github.com/fonttools/fonttools/issues/147.

Replaces all glyph names with ones like "cid12345".

Why do we want to promote CFFs to being CID-keyed?

### CODE REVIEW

File: Lib/fontTools/subset/__init__.py [hunk 1]
```diff
@@ -2466,6 +2466,21 @@ def prune_post_subset(self, options):
 		for subrs in all_subrs:
 			del subrs._used, subrs._old_bias, subrs._new_bias
 
+	if not options.glyph_names:
```
[the remark this answers is not in the document]
Comment (madig) [reply Lib/fontTools/subset/__init__.py]: Does this imply there's nothing to be done for already-CID-keyed fonts?

**Comment by anthrotype** [comment]
the tests' expected results need to be adjusted accordingly.

File: Lib/fontTools/subset/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mashabow) [reply Lib/fontTools/subset/__init__.py]: In the [predefined character collections](https://github.com/adobe-type-tools/cmap-resources) each CID has its own semantics, and some applications such as InDesign use it to provide better typesetting. Changing ROS to Adobe-Identity-0 and/or renumbering CIDs mean to lost these semantics.

[the remark this answers is not in the document]
Comment (behdad) [reply Lib/fontTools/subset/__init__.py]: Though one can argue that dropping glyphnames has the same effect...

**Comment by miguelsousa** [comment]
The changes in this PR are about the `--no-glyph-names` option, so the tests that are failing shouldn't be, I think.

test_desubroutinize_CFF
test_no_hinting_CFF
test_no_hinting_desubroutinize_CFF
test_no_notdef_outline_otf
test_subset_math

**Comment by anthrotype** [comment]
> the tests that are failing shouldn't be

everything in ttx is referenced with glyph names. if you replace things like "a" with "cid00001", of course all the tests asserting something is equal to an expected TTX dump will fail.

**Comment by madig** [comment]
Still need to check if we actually save bytes with this...

**Comment by miguelsousa** [comment]
> everything in ttx is referenced with glyph names. if you replace things like "a" with "cid00001", of course all the tests asserting something is equal to an expected TTX dump will fail.

But none of the failing tests call the `--no-glyph-names` option. Look at [test_no_hinting_CFF](https://github.com/fonttools/fonttools/blob/master/Tests/subset/subset_test.py#L405-L406) for example. If `--no-glyph-names` is NOT being used why should the glyph names in the resulting file be in the form *cidXXXXX*? Am I missing something?

**Comment by anthrotype** [comment]
> none of the failing tests call the --no-glyph-names option

--no-glyph-names is the default:
https://github.com/fonttools/fonttools/blob/2bed5e79d6b88f40945a121b3ab3e0c2d153ddc4/Lib/fontTools/subset/__init__.py#L2758

**Comment by miguelsousa** [comment]
> --no-glyph-names is the default

That explains it, thanks. I guess I was never aware that it was the default because the CFF part of that option only now is being implemented.

Having `--no-glyph-names` as default seems a bit excessive to me, because it will produce significant changes to the font's structure, which will be in addition to the expected subset changes.

**Comment by anthrotype** [comment]
remember that the subsetter was originally designed for webfonts, where one wants to squeeze every bit out of it.

> it will produce significant changes to the font's structure

so is this optimization being discussed here more aggressive than dropping the glyph names (switching to post format 3) in a TrueType-flavored OpenType font, which is what the subsetter has always been doing?

**Comment by miguelsousa** [comment]
> so is this optimization being discussed here more aggressive

no, I see those as the same approach that only differ because of the format. (Using `post` format 3 is actually something I've used in CFF2 fonts). I think I'm just hung up on it being the default, but you've explained why that is. I'll get over it eventually.

**Comment by anthrotype** [comment]
it does look more of a change compared to the other, perhaps because when we drop glyph names from TT font and dump it with ttx, the new glyph names are generated from the cmap and AGLFN, so they are more alike the ones that were dropped. Whereas with cidXXXX, they become just meaningless indexes and look all the same.

**Comment by anthrotype** [comment]
Maybe what we should do is leave the current default behaviour: i.e. drop glyph names from TrueType fonts (from `post` 2 to 3), but keep the glyph names in the `CFF ` table. Unless an explicit `--no-glyph-names` command line option is provided, in which case we also upgrade the `CFF ` font to CID-keyed.
We would need a tri-state option: `None` (default) means the above, `False` means drop glyph names and do the cid trick, `True` means keep post-2 or CFF names. 
The tricky thing is that custom command-line parser which the subsetter is currently using, which handles options differently based on the type of its default value.
If Nikolaus figures out a way to accomplish that, I'll gladly review and merge, tha

**Comment by behdad** [comment]
Why?

**Comment by behdad** [comment]
We can change fonttools to load names from cmap is CID font is custom encoding.  Then you get same behavior as TrueType.

**Comment by anthrotype** [comment]
> We can change fonttools to load names from cmap is CID font is custom encoding.

yeah, that's better

**Comment by miguelsousa** [comment]
Sounds better to me too

**Comment by behdad** [comment]
Volunteers?

**Comment by miguelsousa** [comment]
I volunteer but it'll be a while until I get time to do it. Too much in my queue ATM.

**Comment by anthrotype** [comment]
Wait, I think this is incomplete.
To convert a name-keyed font to a CID-keyed one, I think we also at least add FDArray and FDSelect to the TopDict.. This patch only modifies the charset data (to use an Identity mapping).

@behdad correct?

**Comment by behdad** [comment]
Sounds right.

**Pull request closed without merging**